### PR TITLE
Add environment variable configuration for BatchAsyncContainerExecutor MaxOperationsInDirectModeBatchRequest- #5295

### LINKED ISSUE #5294: BatchAsyncContainerExecutor customize MaxOperationsInDirectModeBatchRequest
Add ability to customize the MaxOperationsInDirectModeBatchRequest used for BatchAsyncContainerExecutor. Prefer ENVIRONMENT based configuration.

### PR DESCRIPTION
**Pull request by kirankumarkolli**
This PR adds the ability to customize the `MaxOperationsInDirectModeBatchRequest` used by `BatchAsyncContainerExecutor` through environment variable configuration, as requested in the issue.

### 🆕 New Features
- **Environment Variable Support**: Added `COSMOS_MAX_OPERATIONS_IN_DIRECT_MODE_BATCH_REQUEST` environment variable
- **Backward Compatibility**: Falls back to existing `Constants.MaxOperationsInDirectModeBatchRequest` when env var is not set
- **Input Validation**: Validates positive integer values and provides clear error messages for invalid inputs

### 📁 Files Added/Modified
1. **`BatchConfiguration.cs`** (New) - Configuration helper class with environment variable support
2. **`BatchAsyncContainerExecutorCache.cs`** - Updated to use configurable value (minimal 1-line change)
3. **`BatchConfigurationTests.cs`** (New) - Comprehensive unit tests
4. **`BatchAsyncContainerExecutorCacheTests.cs`** - Added integration test

### 🔧 Usage Example
```bash
# Set custom batch operation limit
export COSMOS_MAX_OPERATIONS_IN_DIRECT_MODE_BATCH_REQUEST=200

# Cosmos DB client will now use 200 operations per batch instead of default
CosmosClient client = new CosmosClient(connectionString, new CosmosClientOptions 
{ 
    AllowBulkExecution = true 
});
```

### ✅ Implementation Details
- **Environment Variable Name**: `COSMOS_MAX_OPERATIONS_IN_DIRECT_MODE_BATCH_REQUEST`
- **Expected Values**: Positive integers (e.g., "50", "200", "500")
- **Default Behavior**: Uses existing constant value when env var not set
- **Error Handling**: Throws `ArgumentException` with descriptive messages for invalid values

### 🧪 Testing
Added comprehensive test coverage including:
- Default behavior (no environment variable)
- Custom values from environment variable
- Input validation (negative numbers, invalid strings, zero)
- Edge cases (empty strings, large values)
- Integration testing with `BatchAsyncContainerExecutorCache`

### 🛡️ Quality Assurance
- **Minimal Changes**: Only 1 line modified in existing production code
- **No Breaking Changes**: Existing behavior preserved when env var not set
- **Build Verified**: Project builds successfully
- **Backward Compatible**: Existing tests and functionality unaffected

This implementation provides enterprise-ready configurability for batch operations while maintaining the robustness and reliability of the existing codebase.

Fixes #5294.

---

💬 Share your feedback on Copilot coding agent for the chance to win a $200 gift card! Click [here](https://survey.alchemer.com/s3/8343779/Copilot-Coding-agent) to start the survey.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 1]
```diff
@@ -458,116 +376,77 @@ public static bool IsTcpChannelMultiplexingEnabled()
...
+        public static int GetMaxOperationsInDirectModeBatchRequest()
         {
-            return ConfigurationManager
-                    .GetEnvironmentVariable(
-                        variable: ConfigurationManager.BypassQueryParsing,
-                        defaultValue: false);
+            // If caching is disabled (for testing), always read fresh
+            if (isCachingDisabled)
+            {
+                return GetMaxOperationsInDirectModeBatchRequestInternal();
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Convention: Prefix method names with call name for static like ConfigurationManager.GetMaxOperationsInDirectModeBatchRequestInternal

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 2]
```diff
@@ -121,68 +104,29 @@ internal static class ConfigurationManager
...
-        /// Environment variable to override the HTTP/2 PING keep-alive timeout (in seconds).
-        /// If no PONG response is received within this time, the connection is marked dead. Default: 2 seconds.
-        /// </summary>
-        internal static readonly string Http2KeepAlivePingTimeoutInSeconds = "AZURE_COSMOS_HTTP2_KEEPALIVE_PING_TIMEOUT_IN_SECONDS";
-
-        /// <summary>
-        /// Environment variable name to enable deterministic lease-id partition key values for Change Feed lease creation.
-        /// </summary>
-        internal static readonly string ChangeFeedLeaseIdAsPartitionKeyEnabled = "AZURE_COSMOS_CHANGE_FEED_LEASE_ID_AS_PARTITION_KEY_ENABLED";
+        private static bool isCachingDisabled = false;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Use Testing suffix for Properties and methods for only testing related.

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 3]
```diff
@@ -121,68 +104,29 @@ internal static class ConfigurationManager
...
         /// <summary>
-        /// Environment variable to disable the barrier early yield on 429 optimization.
-        /// When set to "false", the Direct transport layer will not yield early on barrier
-        /// throttling and will instead spin until timeout (pre-3.44.0 behavior).
-        /// Default: true (enabled).
+        /// Cached value for the maximum number of operations in a direct mode batch request.
+        /// This is initialized once and reused to avoid repeatedly reading the environment variable.
         /// </summary>
-        internal static readonly string BarrierEarlyYieldOn429Enabled = "AZURE_COSMOS_BARRIER_EARLY_YIELD_ON_429_ENABLED";
+        private static Lazy<int> maxOperationsInDirectModeBatchRequestCached = new Lazy<int>(GetMaxOperationsInDirectModeBatchRequestInternal);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: In case of GetMaxOperationsInDirectModeBatchRequestInternal failign with exceptions it needs to retry again and not cache the results.

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 4]
```diff
@@ -1,92 +1,92 @@
...
-        /// Environment variable name for overriding optimistic direct execution of queries.
-        /// </summary>
-        internal static readonly string OptimisticDirectExecutionEnabled = "AZURE_COSMOS_OPTIMISTIC_DIRECT_EXECUTION_ENABLED";
-
-        /// <summary>
-        /// Environment variable name to disable sending non streaming order by query feature flag to the gateway.
-        /// </summary>
-        internal static readonly string HybridSearchQueryPlanOptimizationDisabled = "AZURE_COSMOS_HYBRID_SEARCH_QUERYPLAN_OPTIMIZATION_DISABLED";
-
+        internal static readonly string CircuitBreakerTimeoutCounterResetWindowInMinutes = "AZURE_COSMOS_PPCB_TIMEOUT_COUNTER_RESET_WINDOW_IN_MINUTES";
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Unrelated changes are showing up, probably CRLF realted revert them

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 5]
```diff
@@ -87,14 +87,14 @@ internal static class ConfigurationManager
...
 
-        /// <summary>
-        /// A read-only string containing the environment variable name for enabling hub region processing for read requests.
-        /// When enabled (default), the SDK attaches a hub region header to read requests that encounter repeated 404/1002
-        /// (ReadSessionNotAvailable) errors, allowing the hub region to process the request directly. When disabled, the
-        /// SDK falls back to the original retry behavior without hub region header attachment.
-        /// </summary>
-        internal static readonly string HubRegionProcessingEnabled = "AZURE_COSMOS_HUB_REGION_PROCESSING_ENABLED";
-
+        /// <summary>
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Acrtoss the file many unrelated changes are showingup, undo them

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs:54]: Unrelated changes undo

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs (hunk 5 above)
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Unrelated changes undo

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs:125]: Unrelated changes undo

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 6]
```diff
@@ -134,7 +134,14 @@ internal static class ConfigurationManager
         /// be removed once binary encoding is enabled by default for both preview
         /// and GA.
         /// </summary>
-        internal static readonly string TcpChannelMultiplexingEnabled = "AZURE_COSMOS_TCP_CHANNEL_MULTIPLEX_ENABLED";
+        internal static readonly string TcpChannelMultiplexingEnabled = "AZURE_COSMOS_TCP_CHANNEL_MULTIPLEX_ENABLED";
+
+        internal static readonly string MaxOperationsInDirectModeBatchRequest = "AZURE_COSMOS_MAX_OPERATIONS_IN_BATCH_REQUEST";
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Unrelated changes undo

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs:308]: Unrelated changes undo

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 7]
```diff
@@ -134,7 +134,14 @@ internal static class ConfigurationManager
...
+
+        private static Lazy<int> maxOperationsInDirectModeBatchRequestCached =
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Ensure that Lazy will not treat exceptions as results and re-evaluates.

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Explore Laxy CTOR options to achive it.

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 8]
```diff
@@ -134,7 +134,14 @@ internal static class ConfigurationManager
...
         /// </summary>
-        internal static readonly string TcpChannelMultiplexingEnabled = "AZURE_COSMOS_TCP_CHANNEL_MULTIPLEX_ENABLED";
+        internal static readonly string TcpChannelMultiplexingEnabled = "AZURE_COSMOS_TCP_CHANNEL_MULTIPLEX_ENABLED";
+
+        internal static readonly string MaxOperationsInDirectModeBatchRequest = "AZURE_COSMOS_MAX_OPERATIONS_IN_BATCH_REQUEST";
+
+        private static Lazy<int> maxOperationsInDirectModeBatchRequestCached =
+            ConfigurationManager.CreateMaxOperationsInDirectModeBatchRequestCache();
+
+        private static bool isBatchRequestCachingDisabledForTesting;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Disable caching is not a requirement.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/ConfigurationManagerBatchTests.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    using System.Threading;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class ConfigurationManagerTests
+    {
+        [TestMethod]
+        public void LazyPublicationOnly_WhenInitializationFails_Retries()
+        {
+            int attempts = 0;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/ConfigurationManagerBatchTests.cs]: Update test to set different ENV variables, initial ones are invalid entries and then check if it refreshes with valid values.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/ConfigurationManagerBatchTests.cs [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+    [TestClass]
+    public class ConfigurationManagerTests
+    {
+        [TestMethod]
+        public void GetMaxOperationsInDirectModeBatchRequest_WhenEnvironmentValueChanges_Retries()
+        {
+            string environmentValue = "invalid";
+            ConfigurationManager.GetMaxOperationsInDirectModeBatchRequestEnvironmentValueForTesting =
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/ConfigurationManagerBatchTests.cs]: Test the original state or method only.

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 11]
```diff
@@ -457,6 +463,48 @@ public static bool IsTcpChannelMultiplexingEnabled()
                         defaultValue: false);
         }
 
+        public static int GetMaxOperationsInDirectModeBatchRequest()
+        {
+            return ConfigurationManager.maxOperationsInDirectModeBatchRequestCached.Value;
+        }
+
+        internal static void ResetMaxOperationsInDirectModeBatchRequestCacheForTesting()
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: Not needed remove it

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/ConfigurationManagerBatchTests.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+            {
+                Assert.ThrowsException<ArgumentException>(
+                    () => ConfigurationManager.GetMaxOperationsInDirectModeBatchRequest());
+
+                Environment.SetEnvironmentVariable(environmentVariableName, "50");
+                Assert.AreEqual(50, ConfigurationManager.GetMaxOperationsInDirectModeBatchRequest());
+            }
+            finally
+            {
+                Environment.SetEnvironmentVariable(environmentVariableName, null);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/ConfigurationManagerBatchTests.cs:31]: Call Reset

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 13]
```diff
@@ -457,6 +463,48 @@ public static bool IsTcpChannelMultiplexingEnabled()
...
+        {
+            ConfigurationManager.maxOperationsInDirectModeBatchRequestCached =
+                ConfigurationManager.CreateMaxOperationsInDirectModeBatchRequestCache();
+        }
+
+        private static Lazy<int> CreateMaxOperationsInDirectModeBatchRequestCache()
+        {
+            return new Lazy<int>(
+                ConfigurationManager.GetMaxOperationsInDirectModeBatchRequestInternal,
+                LazyThreadSafetyMode.PublicationOnly);
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: The two layers being modeled:

- Inner = `ConfigurationManager.maxOperationsInDirectModeBatchRequestCached` → `Lazy<int>(factory, PublicationOnly)`
- Outer = `ContainerCore.lazyBatchExecutor` → `new Lazy<T>(factory)` (default ExecutionAndPublication)

`LazyThreadSafetyMode.PublicationOnly` was chosen for one specific property: it does not cache exceptions. If the value factory throws, the exception propagates but no state is stored, so the next .Value access re-runs the factory. That is exactly what the new unit test asserts:

```
// ConfigurationManagerBatchTests.cs
Environment.SetEnvironmentVariable(name, "invalid");
ConfigurationManager.ResetMaxOperationsInDirectModeBatchRequestCacheForTesting();

Assert.ThrowsException<ArgumentException>(
    () => ConfigurationManager.GetMaxOperationsInDirectModeBatchRequest());

Environment.SetEnvironmentVariable(name, "50");
Assert.AreEqual(50, ConfigurationManager.GetMaxOperationsInDirectModeBatchRequest());  // retries ✔
```

The test passes. But it passes only because it calls ConfigurationManager directly. I feel it might be broken in production codepath. The only production consumer is `BatchAsyncContainerExecutorCache.GetExecutorForContainer(...)`, and that method is reached exclusively through `ContainerCore`:

```
// ContainerCore.cs:34
private readonly Lazy<BatchAsyncContainerExecutor> lazyBatchExecutor;

// ContainerCore.cs:59  — default ctor overload
this.lazyBatchExecutor = new Lazy<BatchAsyncContainerExecutor>(
    () => this.ClientContext.GetExecutorForContainer(this));

// ContainerCore.cs:70
public override BatchAsyncContainerExecutor BatchExecutor => this.lazyBatchExecutor.Value;
```

`new Lazy<T>(Func<T>)` defaults to `LazyThreadSafetyMode.ExecutionAndPublication`, which does cache exceptions, permanently and by design. So the actual runtime sequence is:

1. App sets `AZURE_COSMOS_MAX_OPERATIONS_IN_BATCH_REQUEST`=abc (or 200, or 0).
2. First bulk `CreateItemAsync` touches `container.BatchExecutor`.
3. Outer Lazy runs its factory → inner Lazy factory throws `ArgumentException`.
4. Inner Lazy (PublicationOnly) correctly discards its state — but the outer Lazy captures the exception and stores it.
5. Operator notices the error and fixes the environment variable.
6. Every subsequent call on that same Container object rethrows the identical cached `ArgumentException`, forever. The inner retry logic is never consulted again.

So, the `PublicationOnly` on the inner cache is defeated by the outer `Lazy`.

File: Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs [hunk 14]
```diff
@@ -69,12 +70,12 @@ internal static class ConfigurationManager
         /// A read-only string containing the environment variable name for capturing the consecutive failure count for writes, before triggering per partition
         /// circuit breaker flow. The default value for this interval is 10 consecutive requests within 1 min window.
         /// </summary>
-        internal static readonly string CircuitBreakerConsecutiveFailureCountForWrites = "AZURE_COSMOS_PPCB_CONSECUTIVE_FAILURE_COUNT_FOR_WRITES";
-
-        /// <summary>
-        /// A read-only string containing the environment variable name for capturing the consecutive failure count for writes, before triggering per partition
-        /// circuit breaker flow. The default value for this interval is 5 consecutive requests within 1 min window.
-        /// </summary>
+        internal static readonly string CircuitBreakerConsecutiveFailureCountForWrites = "AZURE_COSMOS_PPCB_CONSECUTIVE_FAILURE_COUNT_FOR_WRITES";
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/Util/ConfigurationManager.cs]: nit: These are unrelated changes.

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs [hunk 15]
```diff
@@ -270,56 +272,56 @@ public async Task<IReadOnlyList<FeedRange>> GetFeedRangesAsync(
...
-                {
-                    feedTokens.Add(new FeedRangeEpk(partitionKeyRange.ToRange()));
-                }
-
-                return feedTokens;
-            }
-            catch (DocumentClientException dce)
-            {
-                throw CosmosExceptionFactory.Create(dce, trace);
+            try
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs]: nit: Whitespace changes ? Looks like unnecessary.

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs [hunk 16]
```diff
@@ -56,7 +56,9 @@ protected ContainerCore(
             this.Scripts = new ScriptsInlineCore(this, this.ClientContext);
             this.cachedUriSegmentWithoutId = this.GetResourceSegmentUriWithoutId();
             this.queryClient = cosmosQueryClient ?? new CosmosQueryClientCore(this.ClientContext, this);
-            this.lazyBatchExecutor = new Lazy<BatchAsyncContainerExecutor>(() => this.ClientContext.GetExecutorForContainer(this));
+            this.lazyBatchExecutor = new Lazy<BatchAsyncContainerExecutor>(
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:59]: I think we should also add a test to validate this e2e.

**Review by kundadebdatta — DISMISSED** [review]
Some nits, and a comment around adding more tests. LGTM otherwise.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/BatchAsyncContainerExecutorCacheTests.cs [hunk 17]
```diff
@@ -79,6 +79,66 @@ public void Null_When_OptionsOff()
...
+                // Assert
+                Assert.IsNotNull(executor);
+                // The executor should be created with the custom max operations value
+                // We verify this indirectly by ensuring the executor was created successfully
+                // The actual value is verified in the ConfigurationManager tests
+            }
+            finally
+            {
+                // Restore original environment variable value
+                Environment.SetEnvironmentVariable(environmentVariableName, originalValue);
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/Batch/BatchAsyncContainerExecutorCacheTests.cs:112]: ConfigurationManager.ResetMaxOperationsInDirectModeBatchRequestCacheForTesting()

**Review by kundadebdatta — DISMISSED** [review]
Fixing the whitespace diffs before merge would be ideal!

File: Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs [hunk 18]
```diff
@@ -270,56 +272,56 @@ public async Task<IReadOnlyList<FeedRange>> GetFeedRangesAsync(
                 trace,
                 cancellationToken);
 
-            try
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Container/ContainerCore.cs:273]: Undo all unrelated changes might be CRLF related, probably checkout from main and redo changes. Do this across all files where ever applicable.

**Review by kundadebdatta — APPROVED** [review]
LGTM Now.

**Review by aavasthy — APPROVED** [review]
